src/main/flight/mixer.c

```diff
@@ -347,14 +350,53 @@ static void applyFlipOverAfterCrashModeToMotors(void)
     }
 }
 
+#ifdef USE_RPM_LIMITER
+static void applyRpmLimiter(mixerRuntime_t *mixer)
+{
+    static float prevError = 0.0f;
+    static float i = 0.0f;
+
+    const float averageRpm = pt1FilterApply(&mixer->averageRpmFilter, getDshotAverageRpm());
+    const float error = averageRpm - mixer->rpmLimiterRpmLimit;
+
+    // PID
+    const float p = error * mixer->rpmLimiterPGain;
+    const float d = (error - prevError) * mixer->rpmLimiterDGain; // rpmLimiterDGain already adjusted for looprate (see mixer_init.c)
+    i += error * mixer->rpmLimiterIGain;                          // rpmLimiterIGain already adjusted for looprate (see mixer_init.c)
+    i = MAX(0.0f, i);
+    float pidOutput = p + i + d;
+
+    // Throttle limit learning
+    if (error > -100.0f && rcCommand[THROTTLE] < rxConfig()->maxcheck) {
+        mixer->rpmLimiterThrottleScale *= 1.0f - 4.8f * pidGetDT();
+    } else if (pidOutput < -0.05f && rcCommand[THROTTLE] >= rxConfig()->maxcheck && !areMotorsSaturated()) { // Throttle accel corresponds with motor accel
+        mixer->rpmLimiterThrottleScale *= 1.0f + 3.2f * pidGetDT();
+    }
+    mixer->rpmLimiterThrottleScale = constrainf(mixer->rpmLimiterThrottleScale, 0.01f, 1.0f);
+    throttle *= mixer->rpmLimiterThrottleScale;
+
+    // Output
+    pidOutput = MAX(0.0f, pidOutput);
+    throttle = constrainf(throttle - pidOutput, 0.0f, 1.0f);
+    prevError = error;
+
+    DEBUG_SET(DEBUG_RPM_LIMITER, 0, lrintf(error));
+    DEBUG_SET(DEBUG_RPM_LIMITER, 1, lrintf(throttle * 100.0f));
```

Comment:
@haslinghuis 
Do we have more debug fields now in 4.5?

Reply:
We have 8

Reply:
@Tdogb lets use all of them! ha =)

My suggestion:
0 - averageRPMSmoothed
1 - mixerRuntime.rpmLimiterExpectedThrottleLimit * 100.0f
2 - throttle * 100.0f - throttle value after adjustments
3 - error
4 - P
5 - I
6 - D
8 - ??
9 - ??

then can use debug element on osd to see first four to have a rough estimation of how good the limiter works.
And PID values for blackbox explorer

Reply:
8?
9?
Only 7 would be valid :)

Reply:
0 - 7 :)

Reply:
Mostly curiosity about it affecting filter performance, or if it changes noise behavior - but that would also be a viable custom mode, or something where we can select two legacy debug modes, I'd like that option (but could totally give that up.  It would also be a solid default logging mode for any of these race builds, especially if it's the default state for a Spec preset (again, mostly selfish motivations for being able to load usable logs anytime somebody has issues at the field)